src/config/mod.rs

```diff
@@ -0,0 +1,195 @@
+    client_name: String,
+    lifespan: u32,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct WsLongLivedAccessTokenResponse {
+    id: u32,
+    #[serde(rename(deserialize = "type"))]
+    command_type: String,
+    success: bool,
+    result: Option<String>,
+}
+
+type Result<T> = std::result::Result<T, Box<dyn error::Error>>;
+
+pub async fn wait_for_token(config: &YamlConfig) -> Result<GetAccessTokenResponse> {
+    let server = Server::http("0.0.0.0:8000").unwrap();
+    let mut get_token_response: Option<GetAccessTokenResponse> = None;
+    println!("Open http://{}/auth/authorize?client_id=http%3A%2F%2Flocalhost%3A8000&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fcallback in your browser", config.ha.host.as_str());
+    // blocks until the next request is received
+    match server.recv() {
+        Ok(request) => {
+            let url = format!("http://localhost:8000{}", request.url());
+            let query_params: HashMap<_, _> = Url::parse(url.as_str())
+                .unwrap()
+                .query_pairs()
+                .into_owned()
+                .collect();
+            match query_params.get("code") {
+                Some(code) => {
+                    request.respond(Response::from_string(
+                        "Halcyon now authenticated to Home Assistant. You can close this page now.",
+                    ))?;
+                    let either = ha_api::get_access_token(config, code.to_string()).await?;
+                    match either {
+                        either::Left(_) => {
+                            get_token_response = None;
+                        }
+                        either::Right(succ) => {
+                            get_token_response = Some(succ);
+                        }
+                    }
+                }
+                None => {
+                    request.respond(
+                        Response::from_string(
+                            "Something went wrong with authenticating Halcyon with Home Assistant",
+                        )
+                        .with_status_code(500),
+                    )?;
+                    get_token_response = None;
+                }
+            }
+        }
+        Err(e) => {
+            println!("error: {}", e);
+        }
+    };
+    get_token_response
+        .map(Ok)
+        .unwrap_or_else(|| Err("Could not retrieve HA access token".into()))
+}
+
+fn start_ws(config: &YamlConfig, access_token: String) -> Result<WsLongLivedAccessTokenResponse> {
+    let ws_url = format!("ws://{}/api/websocket", config.ha.host);
+    let (mut socket, _) = connect(Url::parse(ws_url.as_str()).unwrap()).expect("Can't connect");
+
+    socket.read_message().expect("Error reading message");
+    let req = WsAuthRequest {
+        auth_type: "auth".to_string(),
+        access_token,
+    };
+
+    let req_as_str = Message::Text(serde_json::to_string(&req)?);
+    socket.write_message(req_as_str).unwrap();
```

Comment:
Can we remove this (and any other `.unwrap`

Really it should be dealt with in the application, and throw reasonable errors. If not `.expect` is slightly more acceptable.